### PR TITLE
add scan_index for improving index generation

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
total account state is exceeding available ram on validators. We are also introducing a new storage format.

#### Summary of Changes
Add `scan_index` to storage apis to allow efficiently returning what is necessary to index storages.
This will result in improved performance if we stop mmapping but also just with the hot storage format.

There is a lot of noise in index generation. This change is not worse and may be better just on time today.

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -8629,41 +8627,70 @@ impl AccountsDb {
...
+                        info.lamports,
+                    ),
+                )
+            });
+            self.accounts_index
+                .insert_new_if_missing_into_primary_index(
+                    slot,
+                    storage.approx_stored_count(),
+                    items,
+                )
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:8690]: nit: This is duplicated in the `else` case too. Can it be moved outside of the `if-else`?

File: accounts-db/src/append_vec.rs [hunk 2]
```diff
@@ -573,10 +593,44 @@ impl AppendVec {
         let aligned_data_len = u64_align!(stored_meta.data_len as usize);
         let next_account_offset = start_of_data + aligned_data_len;
         let offset_to_end_of_data = start_of_data + stored_meta.data_len as usize;
+        let stored_size = start_of_data + aligned_data_len - start_offset;
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs]: Wdyt about this to replace all the vars above?

```rust
let stored_size_unaligned = STORE_META_OVERHEAD + stored_meta.data_len as usize;
let stored_size_aligned = u64_align!(stored_size_unaligned);
let offset_to_end_of_data = start_offset + stored_size_unaligned;
let next_account_offset = start_offset + stored_size_aligned;
```

(This could be a separate PR too.)

File: accounts-db/src/append_vec.rs [hunk 3]
```diff
@@ -573,10 +593,44 @@ impl AppendVec {
...
 
         AccountOffsets {
             next_account_offset,
             offset_to_end_of_data,
+            stored_size,
+        }
+    }
+
+    /// iterate over all pubkeys
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs]: nit: Looks like this comment is incorrect/for an older version of `scan_pubkeys()`? Can it be updated?

In particular, can this mention it is used to scan the append vec *for* index generation? We're not scanning the accounts index (duh), but the name is similar. We also use `index` and `offset` somewhat interchangeably in places, so making sure we're not talking about 'index' here in that manner would be helpful IMO.

File: accounts-db/src/append_vec.rs [hunk 4]
```diff
@@ -573,10 +593,44 @@ impl AppendVec {
...
+                // eof
+                break;
+            };
+            let Some((account_meta, _)) = self.get_type::<AccountMeta>(next) else {
+                // eof
+                break;
+            };
+            let next = Self::next_account_offset(offset, stored_meta);
+            if next.offset_to_end_of_data > self.len() {
+                // data doesn't fit, so don't include this pubkey
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs]: nit:
```suggestion
                // data doesn't fit, so don't include it
```

File: accounts-db/src/accounts_db.rs (hunk 1 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:8690]: `items` is different type in each :(
2 different flavors of iter. `insert_new...` takes an `impl Iterator...`

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:8690]: ahhhh bummer ok

**Review by brooksprumo — DISMISSED** [review]
:shipit: 

Code looks good as-is; can you comment with your thoughts on https://github.com/anza-xyz/agave/pull/524/files#r1547868993 too?

File: accounts-db/src/accounts_db.rs [hunk 5]
```diff
@@ -8619,8 +8619,6 @@ impl AccountsDb {
         if accounts.next().is_none() {
             return SlotIndexGenerationInfo::default();
         }
-        let accounts = storage.accounts.account_iter();
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:8645]: 👍

**Review by brooksprumo — DISMISSED** [review]
:shipit:

File: accounts-db/src/accounts_db.rs [hunk 6]
```diff
@@ -8629,41 +8627,70 @@ impl AccountsDb {
...
+                    slot,
+                    storage.approx_stored_count(),
+                    items,
+                )
+        } else {
+            let accounts = storage.accounts.account_iter();
+            let items = accounts.map(|stored_account| {
+                stored_size_alive += stored_account.stored_size();
+                let pubkey = stored_account.pubkey();
+                if secondary {
```
Comment (HaoranYi) [inline accounts-db/src/accounts_db.rs]: If we reach here, `secondary` will be true alwasy?

File: accounts-db/src/accounts_db.rs [hunk 7]
```diff
@@ -8629,41 +8627,70 @@ impl AccountsDb {
...
 
-            (
-                *pubkey,
-                AccountInfo::new(
-                    StorageLocation::AppendVec(store_id, stored_account.offset()), // will never be cached
-                    stored_account.lamports(),
-                ),
-            )
-        });
+                if let Some(amount_to_top_off_rent_this_account) =
```
Comment (HaoranYi) [inline accounts-db/src/accounts_db.rs]: It seems that in the old code, this block of code runs for `!secondary` case too. Should we keep the old behavior?

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: we don't have the fields to calculate this from each account and we don't want to spend the time to get them. This was only stats at startup that were logged out.

File: accounts-db/src/accounts_db.rs (hunk 6 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: yes

File: accounts-db/src/accounts_db.rs (hunk 7 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: Do we also populate `rent_paying_accounts_by_partition`, which will be used for `collect_rent_in_range`? 
Now we don't have any rent paying accounts. And the rent paying accounts vec is empty. So we are safe.

However,  if a bug or attacker caused a rent paying account, then we will break consensus with this change. 

I think we should remove this rent paying code so that we are absolutely consistent with regards to `!seconsary`. WDYT?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs]: Since `IndexInfo` contains the data len, we could use that to calculate the rent exempt amount. And IndexInfo also has the lamports, so we can see if the account's balance meets the rent exempt minimum. 

This check wouldn't be exactly the same as if we had a ReadableAccount, but it'll be strictly a super set (it'll also contain all the executable accounts, which may be a large number). Is that sufficient?

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: i pushed through all the parameters, with a frowny face.

**Review by HaoranYi — APPROVED** [review]
lgtm

**Review by brooksprumo — APPROVED** [review]
:shipit:
